File: BinarySearch/jichon/14233.js
const fs = require("fs");
const input = fs.readFileSync("./input.txt").toString().trim().split("\n");

function available(job, k) {
  let total = 0;
  for (let i = 0; i < job.length; i++) {
    total += k;
    if (total > job[i]) return false;
  }
  return true;
}
// 더 간결..
// const canWork = (k) => job.every((deadline, i) => k * (i + 1) <= deadline);

function binary(job) {
  let [start, end] = [1, job[job.length - 1]];

  while (start <= end) {
    let mid = Math.floor((end + start) / 2);
    available(job, mid) ? (start = mid + 1) : (end = mid - 1);
  }
  return end; // 제일 큰 놈
}

const solution = () => {
  let job = input[1]
    .split(" ")
    .map(Number)
    .sort((a, b) => a - b);
  console.log(binary(job));
};

solution();

File: BinarySearch/chaechon/14233.py
import sys

input = sys.stdin.readline
n = int(input())
deadline = list(map(int, input().split()))
deadline.sort()

start, end = 1, deadline[-1]

while start <= end:
    mid = (start + end) // 2
    valid = True
    
    # 이 mid 값으로 모든 일의 마감기한이 가능한지 확인
    for i in range(n):
        if mid * (i+1) > deadline[i]:
            valid = False
            break
        
    # 가능하면 start = mid + 1 해서 확인
    # 불가능하면 end = mid - 1 해서 확인
    
    if valid == True:
        start = mid + 1
    else:
        end = mid - 1
            
print(end)